File: Project Phase Camp/78-subsets/78-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        self.arr = []
        self.res = []
        
        def backtrack(idx):
            if idx == len(nums):
                temp = self.arr[::]
                self.res.append(temp)
                return
            
            self.arr.append(nums[idx])
            backtrack(idx + 1)
            self.arr.pop()
            backtrack(idx + 1)
            
        backtrack(0)
        return self.res

File: Project Phase Camp/78-subsets/78-subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
//     global result
    const res = [];
    
//     dfs recursive helper
    const dfs = (i, nums, slate) => {
//         base case
        if(i === nums.length){
            let temp = slate.slice();
            res.push(temp);
            return;
        }
//         dfs recursive case
        dfs(i + 1, nums, slate);
        slate.push(nums[i]);
        dfs(i + 1, nums, slate);
        slate.pop();
    }
    dfs(0, nums, []);
    return res;
};